File: valid-anagram/eunhwa99.java
// 시간 복잡도: O(n) - 문자열 길이 n
// 공간 복잡도: O(1) - 알파벳 개수는 26개로 고정되어 있음
class Solution {
    public boolean isAnagram(String s, String t) {
        if(s.length()!=t.length()) return false; // 길이가 다르면 false

        int[] letterCount = new int[26];
        // 각 알파벳의 개수를 세기 위한 배열
        for(int i = 0; i < s.length(); i++){
            letterCount[s.charAt(i) - 'a']++;
            letterCount[t.charAt(i) - 'a']--;
        }

        for(int count : letterCount){
            if(count != 0) return false;
        }
        return true;

    }
}

File: number-of-connected-components-in-an-undirected-graph/eunhwa99.kt
class Solution {
    // Time Complexity: O(n + e)
    // Space Complexity: O(n + e)
    // n : number of nodes
    // e : number of edges
    fun countComponents(n: Int, edges: Array<IntArray>): Int {
        val graph = Array(n) { mutableListOf<Int>() }
        for (edge in edges) { // make graph
            graph[edge[0]].add(edge[1])
            graph[edge[1]].add(edge[0])
        }
        val visited = BooleanArray(n) // visited array
        var count = 0 // number of connected components

        for (i in 0 until n) {
            if (!visited[i]) {
                dfs(i, graph, visited)
                count++
            }
        }
        return count
    }

    fun dfs(node: Int, graph: Array<MutableList<Int>>, visited: BooleanArray) {
        visited[node] = true
        for (neighbor in graph[node]) {
            if (!visited[neighbor]) {
                dfs(neighbor, graph, visited)
            }
        }
    }
}
